spec.md

```diff
@@ -570,13 +559,297 @@ schema being used.
         "bullets": [
           "40 concurrent connections"
         ]
+      },
+      "schemas": {
+        "service_instance": {
+          "create": {
+            "parameters": {
+              "$ref": "osb_v2:///fakeservice/v1#BillingAccountParameters"
+            }
+          },
+          "update": {
+            "parameters": {
+              "$schema": "http://json-schema.org/draft-04/schema#",
+              "type": "object",
+              "propeties": {
+                "some-other-account": {
+                  "description": "Other fake account needed for update.",
+                  "type": "string"
+                },
+                "billing": {
+                  "$ref": "osb_v2:///fakeservice/v1#BillingAccountParameters"
+                }
+              }
+            }
+          }
+        },
+        "service_binding": {
+          "create": {
+            "parameters": {
+              "$ref": "osb_v2:///fakeservice/v1/bindings#AccountParameters"
+            }
+          }
+        }
       }
     }]
   }]
 }
 ```
 
 
+## JSON Schemas
+
+The JSON Schemas endpoint is used to define reusable definitions of JSON Schema
+Objects referenced throughout the [Catalog](#catalog-management). If a Platform
+intends to support JSON Schemas, it MUST examine the response from the catalog
+to determine if the Service Broker is using `$ref` JSON Schemas. The Service
+Broker can signal to the Platform that JSON Schema components are supported by
+using the `osb_v2:` scheme. Components are a way to allow Platforms to only
+fetch sub-schemas.
+
+The root JSON Schema document can be found by getting the JSON Schemas
+endpoint without a `:component_path`.
+
+If the `osb_v2:` scheme is used by the Service Broker, then the Platform MAY
+fetch the component JSON Schema directly using the following convention:
+
+ - Given: `osb_v2:///{:component_path}`, fetch
```

Comment:
Suppose it were a goal to rely on 'pure json schema' facilities in terms of references; motivation for such goal may be using existing JSON schema libraries to fetch/resolve/interpret references. How would the proposal change with such goal in mind?

Reason I wonder is that OSB being an HTTP-based API, ideally there should be no reason why the schema support in OSB shouldn't rely on base JSON schema facilities, fetching schemas via HTTP and using combination of external and internal $refs to achieve its goals.j

Then, the only constraint we may choose to add is that all external refs (and those could be exclusively contained within the /v2/catalog response) point to that very broker's /v2/catalog/schemas endpoint.

That would lead to broker being transitively closed world of schemas.

Yet another way altogether to achieve this feature may be to drop the constraint which prohibits brokers to use external refs in their schemas, and allow brokers to serve schemas in variety of ways (similar to how dashbaord client is handled by the spec). This last option is not my favorite :) but I'd prefer solution which does less customization of JSON schema.

Reply:
This proposal works with existing json schema libraries as demonstrated in the design doc in the description. The only requirement we are really applying is that json schema libraries are not allowed to fetch schemas. They must be provided to the library, and the platform needs a mechanism to work out where to fetch these from. 

Then the broker author must make a choice, either they support `osb_v2` style uris and the fetching of sub-resources, or they don't and the broker expects the platform to get the schema objects from `/v2/schemas`